Add Vercel Project Explorer

MaxLeiter opened the pull request:
## Description

Adds a new Vercel Project Explorer extension that allows you to interact with and explore your projects / deployments.

## Screencast
I can record one in the future, but for now:

<details><summary>Screenshots (click to expand)</summary>
<img width="1000" alt="vercel-project-explorer-1" src="https://user-images.githubusercontent.com/8675906/178161985-02404745-54f9-4abf-af44-0b150765cc88.png">
<img width="1000" alt="vercel-project-explorer-2" src="https://user-images.githubusercontent.com/8675906/178161988-b8462858-9381-493c-b32f-69f7ea4004e0.png">
<img width="1000" alt="vercel-project-explorer-3" src="https://user-images.githubusercontent.com/8675906/178161990-185010cc-efcd-4fd8-87d5-ec927457f7bf.png">
</details>


## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

koinzhang commented:
Hi @MaxLeiter There is something wrong with this [screenshot](https://github.com/raycast/extensions/pull/2106/files#diff-78740a86457e219f9e562e2b4f1f65dddf4b08e63ae425a1ab9ff3e70408c443), the Toast in the bottom left corner

danielsequeira commented:
> Hi @MaxLeiter There is something wrong with this [screenshot](https://github.com/raycast/extensions/pull/2106/files#diff-78740a86457e219f9e562e2b4f1f65dddf4b08e63ae425a1ab9ff3e70408c443), the Toast in the bottom left corner

@MaxLeiter on top of this could you please rename your screenshots in the metadata folder so they all have the same name and a number suffix? To make this easier for you, you should use our new Rayfie tool. To enable it go to advanced tab in preferences:
![Screenshot 2022-06-30 at 18 13 43](https://user-images.githubusercontent.com/12091319/176726877-b2af4926-1653-461a-abb2-21f3cdc6f231.png)

danielsequeira commented:
Hey @MaxLeiter 👋 

Thank you for creating your first Raycast extension.

I had a look at the code and tried the extension myself and I would like to propose to you a few improvements to it.

First I see that you added three different main functionality areas under the same command: Projects, Teams and Account information. I think it would make more sense and also look more aligned with how normally our extensions are built if you had one different command for each different feature. A command to Search Projects, another to Create New Project, one to Switch Team, Create New Team, Display Account information, etc.

The other improvement would be related to the create commands. At the moment the action you have is to open the create form pages in the browser, but it would be amazing if you could actually build the create form in Raycast, as it happens with so many of our native and external extensions. You can check this out in extension like Linear, GitHub or Reminders.

Most importantly regarding these proposed improvements, if you have any questions or need any help we are happy to do it 🙂

MaxLeiter commented:
Thanks for the feedback, @danielsequeira.

I've fixed the screenshots and duplicated them in a `media` folder for the README. 

Regarding the `new project` command, I can build a basic UI but the [Vercel /new page ](https://vercel.com/new) is going to be far more useful and easier than anything Raycast can have, imo. Most of the parameters are documented at https://vercel.com/docs/rest-api#endpoints/projects/create-a-new-project but `gitRepository` (an important one) is just defined as `object`, and I don't really feel like relying on an undocumented API. The other features are all implemented in Raycast, iirc. 

As for multiple commands, I'm happy to make the changes you suggest if the team thinks they're better, but I'm not sure it makes a lot of sense. For example, selecting your Vercel team will change the output of the other commands, which is why I built it into one (so the user is in the same 'context'). You can change your settings and control your Vercel deployments/projects without leaving the Vercel 'app' in Raycast. It feels more natural to me while I've been using it, but 🤷

danielsequeira commented:
Hello again @MaxLeiter 🙂

> Regarding the `new project` command, I can build a basic UI but the [Vercel /new page ](https://vercel.com/new) is going to be far more useful and easier than anything Raycast can provide, imo. Most of the parameters are documented at https://vercel.com/docs/rest-api#endpoints/projects/create-a-new-project but `gitRepository` (an important one) is just defined as `object`, and I don't really feel like relying on an undocumented API. The other features are all implemented in Raycast, iirc.

I was just reading through the API docs you shared and indeed doesn't seem as straightforward as I thought to create this type of form in Raycast. For now I think it is fine to have the create commands opening a link to the corresponding creation pages/forms in vercel. Maybe something you can tackle in a future version of this extension.

> As for multiple commands, I'm happy to make the changes you suggest if the team thinks they're better, but I'm not sure it makes a lot of sense. For example, selecting your Vercel team will change the output of the other commands, which is why I built it into one (so the user is in the same 'context'). You can change your settings and control your Vercel deployments/projects without leaving the Vercel 'app' in Raycast. It feels more natural to me while I've been using it, but 🤷. One example: You can go fairly far into a projects page: should adding a new environment variable be it's own command? or is that a sub-command of 'View Vercel projects'? Where do we draw the line on a sub command versus it's own command?

We do believe is a lot better to have different commands for different types of functionality, this will allow you to assign them different aliases, shortcuts, etc. And even if they are related to each other, they do serve different purposes and that is how all our extensions are built (native and third-party). We also deeply believe that this offers a better UI and UX inside Raycast.

You do make a good point regarding Teams and switching between them, we have a few extensions with the same functionality and for that we recommend using a dropdown on the top right corner. As it is done in the GitHub extension to switch between open issues of different repositories:
![Raycast 2022-07-04 at 14 32 23](https://user-images.githubusercontent.com/12091319/177155827-d4023bd9-6f3c-4ba5-8fcf-97a2155265f7.png)

If you implement the above you don't need the "Switch team..." as a separate command.

On this subject your [project screen](https://user-images.githubusercontent.com/8675906/176060778-dcddb658-8aa5-45bb-b0ef-5b4bcde64ee6.png) could also be improved. Instead of listening all the actions that could be done within a project, considering the way Raycast works and what we offer in our API, it would make more sense and improve the experience if those things were defined within the Action Panel. As mentioned before, this would also allow for each one of these actions to have its own shortcut. This would also allow you to display some information and details about the project. As we do for example in the GitHub extension for Pull Requests:
![Raycast 2022-07-04 at 14 44 56](https://user-images.githubusercontent.com/12091319/177157795-2bc9a5cc-70d4-44d8-83b4-f873fbabc108.png)

One last thing the team mentioned was the fact that we do already have a [Vercel extension](https://www.raycast.com/matt/vercast) which offers users the possibility to list and search their Vercel deployments. Instead of having your extension and code being a modified fork of this one, as you mentioned in your README. Have you consider having it as an addition to it?

MaxLeiter commented:
Hey @danielsequeira!

> Maybe something you can tackle in a future version of this extension.

Agreed. I removed it for now as it was out of place with the refactored commands, and ultimately not very useful.

> We do believe is a lot better to have different commands for different types of functionality, this will allow you to assign them different aliases, shortcuts, etc... for [switching teams] we recommend using a dropdown on the top right corner.

This has been done.

> One last thing the team mentioned was the fact that we do already have a [Vercel extension](https://www.raycast.com/matt/vercast) which offers users the possibility to list and search their Vercel deployments. Instead of having your extension and code being a modified fork of this one, as you mentioned in your README. Have you consider having it as an addition to it?

@gleich and I have talked in the past and neither of us had the motivation / time to properly merge the extensions. This is basically a complete rewrite/refactor, and is no longer really a fork. I don't mind this extension replacing that one in the store, but I'm fine with both existing and letting the user decide which best suits them.


New screenshots:
<img width="1000" alt="vercel-project-explorer 2022-07-08 at 22 39 11" src="https://user-images.githubusercontent.com/8675906/178093439-b8dbd281-8c1f-4b2a-b6e9-38619af2cc84.png">
<img width="1000" alt="vercel-project-explorer 2022-07-08 at 22 38 51" src="https://user-images.githubusercontent.com/8675906/178093443-3080ffc6-7845-402a-90cc-cfca79ec71cc.png">
<img width="1000" alt="vercel-project-explorer 2022-07-08 at 22 38 03" src="https://user-images.githubusercontent.com/8675906/178093444-b6a4b612-9270-4b47-a511-d6e097ee8554.png">
<img width="1000" alt="vercel-project-explorer 2022-07-08 at 22 37 42" src="https://user-images.githubusercontent.com/8675906/178093445-084231cf-3eee-47ec-9118-53dec53dcd5f.png">

danielsequeira commented:
Hey @MaxLeiter,

Thank you so much for the changes. I just had another go at it and found a couple of issues.

1. In "Search Deployments" command if I'm in an account without any deployments, the app keeps loading forever. I would expect the same behaviour as in "Search Projects" where the "No Results" message would appear.
![vercel-project-explorer 2022-07-11 at 13 32 07](https://user-images.githubusercontent.com/12091319/178256455-dae6df1f-d544-4d47-8a7e-88e402ec81cd.png)
![vercel-project-explorer 2022-07-11 at 13 32 59](https://user-images.githubusercontent.com/12091319/178256570-bc0ab3b3-7786-4b28-9888-1e6093449735.png)

2. In "View User Information" I would suggest to remove the "Account information" title, since it doesn't really add any value and it is not coherent with the name of the command as well.
![vercel-project-explorer 2022-07-11 at 13 32 50](https://user-images.githubusercontent.com/12091319/178256945-a178c0a5-1894-470c-b248-f0e974e6eaeb.png)

3. When switching between accounts in the dropdown, there is some weird spacing in my "Raycast" option due to the account icon/avatar being missing, does the same happen to you?
![vercel-project-explorer 2022-07-11 at 13 33 11](https://user-images.githubusercontent.com/12091319/178257328-a5de9bcc-4fda-4813-9c42-dfd08afe45e3.png)

4. You didn't seem to address my comments regarding your project screen:
> On this subject your [project screen](https://user-images.githubusercontent.com/8675906/176060778-dcddb658-8aa5-45bb-b0ef-5b4bcde64ee6.png) could also be improved. Instead of listening all the actions that could be done within a project, considering the way Raycast works and what we offer in our API, it would make more sense and improve the experience if those things were defined within the Action Panel. As mentioned before, this would also allow for each one of these actions to have its own shortcut. This would also allow you to display some information and details about the project. As we do for example in the GitHub extension for Pull Requests:

5. In my opinion I would remove the command "Modify Environment Variables" from the root and keep it as an action of the projects. Also this doesn't really work properly because it doesn't the user to switch between accounts, which for me results in being stuck in my personal account where I don't have any projects.
![vercel-project-explorer 2022-07-11 at 13 34 11](https://user-images.githubusercontent.com/12091319/178258015-a3b88bc7-6afb-4226-af26-e206ffb1f7a8.png)

MaxLeiter commented:
Hey @danielsequeira, *now* I think I've addressed all your comments 😅 

1. 
> In "Search Deployments" command if I'm in an account without any deployments, the app keeps loading forever. I would expect the same behaviour as in "Search Projects" where the "No Results" message would appear.

Fixed, and took the opportunity to switch to the useFetch hook

2.  
> I would suggest to remove the "Account information" title

I removed this command, it's not very useful except when debugging.

3. 
>  there is some weird spacing in my "Raycast" option...

Fixed by removing the attempt at loading profile pics, the API isn't public and it can return png or svg, which I don't want to have to detect

4.
> You didn't seem to address my comments regarding your project screen

I think I've fixed this. I removed the inspect project page and replaced it with these actions:

<img width="747" alt="image" src="https://user-images.githubusercontent.com/8675906/180949164-f9c95a06-f6e3-4523-a5b9-6a7833e24233.png">

5. 
> In my opinion I would remove the command "Modify Environment Variables" from the root and keep it as an action of the projects. 

Done.

danielsequeira commented:
@MaxLeiter thank you so much for the changes, we really appreciate it 🙂 

It seems that we have a few formatting issues that broke one of the pipelines though.

MaxLeiter commented:
@danielsequeira oops, I fixed the formatting.

`Open on Vercel` opens the project/deployments vercel.com page. `Visit Most Recent Deployment` opens the deployed website. Anything that starts with `Inspect` opens a Raycast view. I'm open to suggestions on the icons / text to make the actions more clear

gleich commented:
Hey @MaxLeiter! Love the work you've done with your extension. I'd be interested in having another convo about merging the extensions. I'll shoot you an email with more information on Thursday.

danielsequeira commented:
@MaxLeiter the "Open on Vercel" action is not working for me, either it opens the browser with the URL: https://undefined/ on the Search Projects command and on the Search Deployments it opens the browser with the URL starting as: https://https//vercel.com/... which is also invalid.

On another note @gleich and @MaxLeiter I will hold on my review until you guys discuss the merging of both extensions, that would be amazing

MaxLeiter commented:
@danielsequeira I fixed up all the inspect / open links.

danielsequeira commented:
Thanks @MaxLeiter 🙏 

I did notice another small issue though. When I trigger the action to show details of a project or deployment. While fetching the data, an error message appears in the action bar, mentioning that the fetch failed.

https://user-images.githubusercontent.com/12091319/181740288-be171c86-0f1b-4276-8c90-4b20efd4fd8b.mp4

MaxLeiter commented:
@danielsequeira I'm not currently in a Vercel team but I believe I pushed a fix if you can verify it works. I forgot to add a `teamId` query to an API call.

gleich commented:
Once all the issues have been resolved I am cool with this extension replacing mine

danielsequeira commented:
Thanks for fixing it @MaxLeiter, I'm not experiencing it anymore.

However I did find another small issue. I have two teams in my vercel account: danielsequeira and Raycast. danielsequeira is selected by default, but it is empty. If I go into search projects and change teams to Raycast, I'm able to see the Raycast projects, but if in one of them I trigger the action "Search deployments" it will go to the deployments screen but with danielsequeira team selected, which means it won't find the deployments of the project. Video below showcasing this:

https://user-images.githubusercontent.com/12091319/182330986-65f7a0f3-43e0-4ff6-9888-bce3bc3b7d0b.mp4

danielsequeira commented:
> Once all the issues have been resolved I am cool with this extension replacing mine

@gleich unfortunately this is not that easy since your extension has more than 1500 installs and I imagine a lot of people are using it already. That is why my first proposal was for @MaxLeiter to just add his changes to the current extension code. However I also get that this has its challenges. For now we will need to keep and support both. We will look into what we can do here, maybe a deprecated warning pointing to this new one could work.

leerob commented:
I'd love to have a single extension to point folks to - especially given how popular the current one it (but it is currently not working) - so very excited about this one!

MaxLeiter commented:
I'm not entirely sure what happened to the commits but I fixed your problem and potentially similar ones by changing some logic @danielsequeira. 

For the extensions in the store, would I just need to `rm -rf` the existing one, rename this one, and update the package.json? Or is there more I'd need to do to replace the existing one?

danielsequeira commented:
@MaxLeiter I just checked with the team and we think the best way to go about this one, would be to put this new code in the current location of the other vercel extension. This way we would treat it as a big refactor of the current extension.

MaxLeiter commented:
@danielsequeira I believe that's complete 👍

leerob commented on extensions/vercast/README.md:
```diff
@@ -1,5 +1,27 @@
+## Contributing
+The [Vercel REST API docs](https://vercel.com/docs/rest-api) and [Raycast developer docs](https://developers.raycast.com/) are good places to start. Feel free to open an issue or [ping me on Twitter](https://twitter.com/max_leiter) if you want to help out.
+
+This is a (heavily) modified fork of the [vercast](https://github.com/raycast/extensions/tree/main/extensions/vercast) extension.
```
```suggestion
```

leerob commented on extensions/vercast/package.json:
```diff
@@ -1,54 +1,55 @@
   "name": "vercast",
-  "title": "Vercel Deployments",
-  "description": "View your Vercel deployments",
+  "title": "Vercast",
```
```suggestion
  "title": "Vercel Deployments",
```

MaxLeiter replied:
I was hesitant about this because it's not really just your deployments

leerob commented on extensions/vercast/package.json:
```diff
@@ -1,54 +1,55 @@
+  "description": "View and modify your Vercel projects and deployments.",
   "icon": "icon.png",
-  "author": "matt",
+  "author": "MaxLeiter",
```
I wonder if you can have multiple authors

MaxLeiter replied:
The schema specifically mentions `User currently contributing the most to the extension.`, so I don't think you can yet. It definitely should be updated to include @gleich when possible, though.

danielsequeira replied:
You can only have one author, but you can add others users to `contributors` property

leerob commented on extensions/vercast/src/hooks/use-vercel-info.ts:
```diff
@@ -0,0 +1,73 @@
+        const [fetchedUser, fetchedTeams] = await Promise.all([fetchUser(), fetchTeams()]);
+
+        // // add avatar to fetched teams
+        // const fetchedTeamsWithAvatar = await Promise.all(fetchedTeams.map(async (team) => {
```
Do we still want this? Or remove?

MaxLeiter replied:
Removed for now, it requires supporting multiple image formats + the API isn't publicly documented

danielsequeira commented:
@MaxLeiter seems that we are missing a CHANGELOG entry

MaxLeiter commented:
I've added matt to the contributors + added a basic changelog

danielsequeira commented on extensions/vercast/package.json:
```diff
@@ -2,53 +2,57 @@
   "$schema": "https://www.raycast.com/schemas/extension.json",
   "name": "vercast",
   "title": "Vercel Deployments",
```
I just checked with the team and it is fine to change this to something else, since the extension now does more than deployments. Should we call it "Vercel"?
```suggestion
  "title": "Vercel",
```
